File: python/src/shadow_mode/comparator.py
# SPDX-License-Identifier: BSL-1.1
# Copyright (c) 2026 MuVeraAI Corporation

"""ShadowComparator — compare a shadow recommendation to an actual decision.

Comparison is pure computation — no external calls, no side effects, no state.
The comparator operates on structured dict outputs and computes:

- Agreement: whether the core decision matched
- Deviation score: quantified difference in [0.0, 1.0]
- Individual deviations: field-level diffs with path, shadow value, actual value
- Risk level: qualitative risk derived from deviation magnitude and field importance

High-priority fields (configurable) carry more weight in deviation scoring.
"""

from __future__ import annotations

from typing import Any

from .types import (
    ActualDecision,
    AgreementLevel,
    ComparisonResult,
    Deviation,
    RiskLevel,
    ShadowDecision,
)


# Default set of field paths considered high-priority for risk assessment.
# Differences in these fields are weighted more heavily in deviation scoring.
DEFAULT_HIGH_PRIORITY_FIELDS: frozenset[str] = frozenset(
    {
        "action",
        "decision",
        "approved",
        "blocked",
        "result",
        "status",
    }
)


class ShadowComparator:
    """Compares a shadow decision to an actual production decision.

    Produces a ``ComparisonResult`` for each pair. The comparison is
    synchronous and purely functional — it inspects the output dicts,
    computes field-level diffs, and calculates a deviation score.

    Args:
        high_priority_fields: Set of top-level field names that are treated
            as high-priority when computing the deviation score and risk level.
            Defaults to ``DEFAULT_HIGH_PRIORITY_FIELDS``.
        agreement_threshold: Deviation score at or below which the comparison
            is considered agreed. Defaults to 0.0 (exact match required for
            full agreement; partial agreement allows up to this threshold).

    Example::

        comparator = ShadowComparator()
        result = comparator.compare(shadow_decision, actual_decision)
        print(result.agreed, result.deviation_score, result.risk_level)
    """

    def __init__(
        self,
        high_priority_fields: frozenset[str] | None = None,
        agreement_threshold: float = 0.1,
    ) -> None:
        if not 0.0 <= agreement_threshold <= 1.0:
            raise ValueError("agreement_threshold must be in [0.0, 1.0].")
        self._high_priority_fields = high_priority_fields or DEFAULT_HIGH_PRIORITY_FIELDS
        self._agreement_threshold = agreement_threshold

    def compare(
        self,
        shadow: ShadowDecision,
        actual: ActualDecision,
    ) -> ComparisonResult:
        """Compare shadow output to actual output and return a scored result.

        Args:
            shadow: The shadow agent's decision (what it would have done).
            actual: The production agent's actual decision.

        Returns:
            ``ComparisonResult`` with agreement, deviation score, field-level
            diffs, and risk level.

        Raises:
            ValueError: If ``shadow.decision_id != actual.decision_id``.
        """
        if shadow.decision_id != actual.decision_id:
            raise ValueError(
                f"decision_id mismatch: shadow={shadow.decision_id!r} "
                f"actual={actual.decision_id!r}. Both must share the same ID."
            )

        deviations = _find_deviations(shadow.output, actual.output)
        deviation_score = _compute_deviation_score(deviations, self._high_priority_fields)
        agreed = deviation_score <= self._agreement_threshold
        agreement_level = _classify_agreement(deviation_score, self._agreement_threshold)
        risk_level = _assess_risk(deviations, deviation_score, self._high_priority_fields)

        return ComparisonResult(
            decision_id=shadow.decision_id,
            agreed=agreed,
            agreement_level=agreement_level,
            deviation_score=round(deviation_score, 6),
            deviations=deviations,
            risk_level=risk_level,
        )


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------


def _find_deviations(
    shadow_output: dict[str, Any],
    actual_output: dict[str, Any],
    path_prefix: str = "",
) -> list[Deviation]:
    """Recursively find field-level deviations between two output dicts.

    Args:
        shadow_output: Shadow agent output (or nested sub-dict).
        actual_output: Actual agent output (or nested sub-dict).
        path_prefix: Dot-separated prefix for nested field paths.

    Returns:
        List of ``Deviation`` objects, one per differing field.
    """
    deviations: list[Deviation] = []
    all_keys = set(shadow_output.keys()) | set(actual_output.keys())

    for key in sorted(all_keys):
        field_path = f"{path_prefix}.{key}" if path_prefix else key
        shadow_value = shadow_output.get(key)
        actual_value = actual_output.get(key)

        if key not in shadow_output:
            deviations.append(
                Deviation(
                    field_path=field_path,
                    shadow_value=None,
                    actual_value=actual_value,
                    description=f"Field '{field_path}' present in actual but missing from shadow.",
                )
            )
        elif key not in actual_output:
            deviations.append(
                Deviation(
                    field_path=field_path,
                    shadow_value=shadow_value,
                    actual_value=None,
                    description=f"Field '{field_path}' present in shadow but missing from actual.",
                )
            )
        elif isinstance(shadow_value, dict) and isinstance(actual_value, dict):
            # Recurse into nested dicts.
            nested = _find_deviations(shadow_value, actual_value, path_prefix=field_path)
            deviations.extend(nested)
        elif shadow_value != actual_value:
            deviations.append(
                Deviation(
                    field_path=field_path,
                    shadow_value=shadow_value,
                    actual_value=actual_value,
                    description=(
                        f"Field '{field_path}' differs: "
                        f"shadow={shadow_value!r}, actual={actual_value!r}."
                    ),
                )
            )

    return deviations


def _compute_deviation_score(
    deviations: list[Deviation],
    high_priority_fields: frozenset[str],
) -> float:
    """Compute a normalised deviation score in [0.0, 1.0].

    High-priority field deviations are weighted at 2x compared to other fields.
    Score is capped at 1.0 regardless of the number of deviations.

    Args:
        deviations: List of field-level deviations.
        high_priority_fields: Set of top-level field names with higher weight.

    Returns:
        Float in [0.0, 1.0]. 0.0 = no deviations, 1.0 = maximum deviation.
    """
    if not deviations:
        return 0.0

    total_weight = 0.0
    for deviation in deviations:
        top_level_field = deviation.field_path.split(".")[0]
        weight = 2.0 if top_level_field in high_priority_fields else 1.0
        total_weight += weight

    # Normalise against expected weight for the number of deviations.
    # A single high-priority deviation in a 1-field output = score of 1.0.
    normaliser = max(total_weight, 1.0)
    raw_score = total_weight / normaliser

    return min(raw_score, 1.0)


def _classify_agreement(
    deviation_score: float,
    threshold: float,
) -> AgreementLevel:
    """Map a deviation score to a qualitative agreement level.

    Args:
        deviation_score: Float in [0.0, 1.0].
        threshold: Maximum score for full agreement.

    Returns:
        ``AgreementLevel.FULL``, ``PARTIAL``, or ``NONE``.
    """
    if deviation_score == 0.0:
        return AgreementLevel.FULL
    if deviation_score <= threshold:
        return AgreementLevel.PARTIAL
    return AgreementLevel.NONE


def _assess_risk(
    deviations: list[Deviation],
    deviation_score: float,
    high_priority_fields: frozenset[str],
) -> RiskLevel:
    """Assess risk level based on deviation score and high-priority field hits.

    Args:
        deviations: List of field-level deviations.
        deviation_score: Overall deviation score.
        high_priority_fields: High-priority field names.

    Returns:
        ``RiskLevel.LOW``, ``MEDIUM``, or ``HIGH``.
    """
    has_high_priority_deviation = any(
        deviation.field_path.split(".")[0] in high_priority_fields
        for deviation in deviations
    )

    if has_high_priority_deviation or deviation_score >= 0.5:
        return RiskLevel.HIGH
    if deviation_score > 0.0:
        return RiskLevel.MEDIUM
    return RiskLevel.LOW

File: typescript/src/comparator.ts
// SPDX-License-Identifier: BSL-1.1
// Copyright (c) 2026 MuVeraAI Corporation

/**
 * ShadowComparator — compare a shadow recommendation to an actual decision.
 *
 * Comparison is pure computation — no external calls, no side effects, no state.
 * Mirrors the Python ShadowComparator in python/src/shadow_mode/comparator.py.
 */

import type {
  ActualDecision,
  AgreementLevel,
  ComparisonResult,
  Deviation,
  RiskLevel,
  ShadowDecision,
} from "./types.js";

/** Default set of top-level field names treated as high-priority for risk assessment. */
const DEFAULT_HIGH_PRIORITY_FIELDS: ReadonlySet<string> = new Set([
  "action",
  "decision",
  "approved",
  "blocked",
  "result",
  "status",
]);

/**
 * Compares shadow decisions to actual decisions and produces ComparisonResult objects.
 *
 * @example
 * ```typescript
 * const comparator = new ShadowComparator();
 * const result = comparator.compare(shadowDecision, actualDecision);
 * console.log(result.agreed, result.deviationScore, result.riskLevel);
 * ```
 */
export class ShadowComparator {
  private readonly highPriorityFields: ReadonlySet<string>;
  private readonly agreementThreshold: number;

  /**
   * @param highPriorityFields - Top-level field names weighted more heavily
   *   in deviation scoring. Defaults to DEFAULT_HIGH_PRIORITY_FIELDS.
   * @param agreementThreshold - Deviation score at or below which the comparison
   *   is considered agreed. Defaults to 0.1.
   */
  constructor(
    highPriorityFields?: ReadonlySet<string>,
    agreementThreshold = 0.1
  ) {
    if (agreementThreshold < 0 || agreementThreshold > 1) {
      throw new RangeError("agreementThreshold must be in [0.0, 1.0].");
    }
    this.highPriorityFields = highPriorityFields ?? DEFAULT_HIGH_PRIORITY_FIELDS;
    this.agreementThreshold = agreementThreshold;
  }

  /**
   * Compare shadow output to actual output and return a scored result.
   *
   * @param shadow - The shadow agent's decision (what it would have done).
   * @param actual - The production agent's actual decision.
   * @returns ComparisonResult with agreement, deviation score, field-level diffs,
   *   and risk level.
   * @throws Error if shadow.decisionId !== actual.decisionId.
   */
  compare(shadow: ShadowDecision, actual: ActualDecision): ComparisonResult {
    if (shadow.decisionId !== actual.decisionId) {
      throw new Error(
        `decisionId mismatch: shadow="${shadow.decisionId}" actual="${actual.decisionId}". ` +
          "Both must share the same ID."
      );
    }

    const deviations = findDeviations(shadow.output, actual.output);
    const deviationScore = computeDeviationScore(deviations, this.highPriorityFields);
    const agreed = deviationScore <= this.agreementThreshold;
    const agreementLevel = classifyAgreement(deviationScore, this.agreementThreshold);
    const riskLevel = assessRisk(deviations, deviationScore, this.highPriorityFields);

    return {
      decisionId: shadow.decisionId,
      agreed,
      agreementLevel,
      deviationScore: roundTo6(deviationScore),
      deviations,
      riskLevel,
    };
  }
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

function findDeviations(
  shadowOutput: Record<string, unknown>,
  actualOutput: Record<string, unknown>,
  pathPrefix = ""
): Deviation[] {
  const deviations: Deviation[] = [];
  const allKeys = new Set([...Object.keys(shadowOutput), ...Object.keys(actualOutput)]);

  for (const key of [...allKeys].sort()) {
    const fieldPath = pathPrefix ? `${pathPrefix}.${key}` : key;
    const shadowValue = shadowOutput[key];
    const actualValue = actualOutput[key];

    if (!(key in shadowOutput)) {
      deviations.push({
        fieldPath,
        shadowValue: undefined,
        actualValue,
        description: `Field '${fieldPath}' present in actual but missing from shadow.`,
      });
    } else if (!(key in actualOutput)) {
      deviations.push({
        fieldPath,
        shadowValue,
        actualValue: undefined,
        description: `Field '${fieldPath}' present in shadow but missing from actual.`,
      });
    } else if (isPlainObject(shadowValue) && isPlainObject(actualValue)) {
      const nested = findDeviations(
        shadowValue as Record<string, unknown>,
        actualValue as Record<string, unknown>,
        fieldPath
      );
      deviations.push(...nested);
    } else if (!deepEqual(shadowValue, actualValue)) {
      deviations.push({
        fieldPath,
        shadowValue,
        actualValue,
        description:
          `Field '${fieldPath}' differs: ` +
          `shadow=${JSON.stringify(shadowValue)}, actual=${JSON.stringify(actualValue)}.`,
      });
    }
  }

  return deviations;
}

function computeDeviationScore(
  deviations: Deviation[],
  highPriorityFields: ReadonlySet<string>
): number {
  if (deviations.length === 0) return 0;

  let totalWeight = 0;
  for (const deviation of deviations) {
    const topLevelField = deviation.fieldPath.split(".")[0] ?? deviation.fieldPath;
    const weight = highPriorityFields.has(topLevelField) ? 2.0 : 1.0;
    totalWeight += weight;
  }

  const normaliser = Math.max(totalWeight, 1);
  return Math.min(totalWeight / normaliser, 1.0);
}

function classifyAgreement(deviationScore: number, threshold: number): AgreementLevel {
  if (deviationScore === 0) return "full";
  if (deviationScore <= threshold) return "partial";
  return "none";
}

function assessRisk(
  deviations: Deviation[],
  deviationScore: number,
  highPriorityFields: ReadonlySet<string>
): RiskLevel {
  const hasHighPriorityDeviation = deviations.some((d) =>
    highPriorityFields.has(d.fieldPath.split(".")[0] ?? d.fieldPath)
  );

  if (hasHighPriorityDeviation || deviationScore >= 0.5) return "high";
  if (deviationScore > 0) return "medium";
  return "low";
}

function isPlainObject(value: unknown): boolean {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function deepEqual(a: unknown, b: unknown): boolean {
  return JSON.stringify(a) === JSON.stringify(b);
}

function roundTo6(value: number): number {
  return Math.round(value * 1_000_000) / 1_000_000;
}
